pynicotine/portmapper.py

```diff
@@ -402,17 +483,64 @@ class PortMapper:
     LEASE_DURATION = 43200    # 12 hours
 
     def __init__(self):
+
+        self._active_implementation = None
         self._timer = None
+        self._natpmp = NATPMP()
         self._upnp = UPnP()
 
     def set_port(self, port, local_ip_address):
+        self._natpmp.set_port(port, local_ip_address)
         self._upnp.set_port(port, local_ip_address)
 
     def _add_port_mapping(self):
-        self._upnp.add_port_mapping(self.LEASE_DURATION)
+
+        log.add_debug("Creating Port Mapping rule...")
+
+        try:
+            try:
+                self._active_implementation = self._natpmp
+                self._natpmp.add_port_mapping(self.LEASE_DURATION)
+
+            except RuntimeError as error:
+                log.add_debug("NAT-PMP not available, falling back to UPnP: %s", error)
```

Comment:
Are we not starting with UPnP and falling back to NAT-PMP, whereas this is the other way around here?

Reply:
I believe NAT-PMP is available on most routers in addition to UPnP, and it's simpler and faster than UPnP, which is why I opted to try it first.

Reply:
That does make some logical sense.

My only concern might be that if users have been used to seeing an entry in their UPnP configuration, then this would potentially change that behaviour, albeit it is inconsequential once the port has been established in most cases.

Reply:
We can test this behavior in master for a while, and see how it works.